### PR TITLE
Add solution for Decode the Message

### LINKED ISSUE #32: Decode the Message
## Issue Overview
Implement a solution for the "Decode the Message" LeetCode problem to practice string substitution and mapping in Rust.

## Feature Details
Create a Rust implementation for the problem: https://leetcode.com/problems/decode-the-message/

Problem summary:
You are given a string `key` representing a substitution cipher key, and a string `message`. Decode the `message` using the key according to the specified rules and return the decoded message.

Constraints:
- 26 <= key.length <= 2000
- key consists of lowercase English letters and ' '.
- key contains every letter in the English alphabet ('a' to 'z') at least once.
- 1 <= message.length <= 2000
- message consists of lowercase English letters and ' '.

The solution should:
- Create and utilize an efficient character mapping system
- Handle string transformation in an idiomatic Rust way
- Include proper error handling for invalid inputs
- Feature comprehensive tests

## Expected Outcome
A well-documented Rust implementation that correctly decodes the message with clean, efficient code, proper error handling, and thorough test coverage.

### PR DESCRIPTION
**Pull request by zmrdltl**
Related Issues
closes #32 

PR Description
Adds a solution for the LeetCode "Decode the Message" problem.

Implementation details:
- Create a substitution cipher decoder that maps characters from key to alphabet sequence
- Preserve the first occurrence order of each character in the key
- Process each character in the message and replace with mapped alphabet letters
- Use type-safe conversions with proper error handling
- Key features:
  - Sequential mapping of unique key characters to alphabet (a-z)
  - Space character preservation in messages
  - HashSet for efficient duplicate removal
  - Input validation for key and message constraints

The implementation prioritizes safety by avoiding unsafe type casts and
handling all potential error conditions explicitly. The solution uses a
combination of standard collections (HashSet, HashMap, Vec) to efficiently
track character mappings while preserving original insertion order.

Comprehensive tests verify both the main functionality and edge cases,
including input validation for length constraints and character content.

Special Notes
The approach avoids unsafe operations like unwrap() and as casting, preferring
explicit error handling with Result and proper type conversions.

### CODE REVIEW

File: src/encoding_decoding/decode_the_message.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        .chars()
+        .any(|c| c != ' ' && (!c.is_ascii_alphabetic() || !c.is_ascii_lowercase()))
+    {
+        return Err("Message must contain only lowercase letters");
+    }
+
+    let mut key_chars = Vec::new();
+    let mut seen = HashSet::new();
+
+    for c in key.chars() {
```
Comment (kimhanbeom) [inline src/encoding_decoding/decode_the_message.rs:53]: HashMap생성 for문에서 모든 처리가 가능해 보이므로 key_chars와 seen는 굳이 생성할 필요가 없어 보입니다.

File: src/encoding_decoding/decode_the_message.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        if c != ' ' && !seen.contains(&c) {
+            seen.insert(c);
+            key_chars.push(c);
+        }
+    }
+
+    let mut key_chars_map = HashMap::new();
+    for (i, &c) in key_chars.iter().enumerate() {
```
Comment (kimhanbeom) [inline src/encoding_decoding/decode_the_message.rs:61]: LOWERCASE_A를 mutable 변수로 변경하고 이 값을 하나씩 증가시키면 굳이 `enumerate`를 호출할 필요가 없어 보입니다.

**Review by kimhanbeom — CHANGES_REQUESTED** [review]

File: src/encoding_decoding/decode_the_message.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    key.chars()
+        .filter(|&c| c != ' ')
+        .scan(0, |index, c| {
+            if *index < alphabet.len() && !key_chars_map.contains_key(&c) {
+                key_chars_map.insert(c, alphabet[*index]);
+                *index += 1;
+            }
+            Some(())
+        })
+        .count();
```
Comment (kimhanbeom) [inline src/encoding_decoding/decode_the_message.rs:60]: `HashMap` 생성을 위해 의미없는 `count`까지 호출되고 있는데 이런 경우는 for문을 사용하는 것이 더 나아 보입니다.

File: src/encoding_decoding/decode_the_message.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    let decoded_message = message
+        .chars()
+        .map(|c| {
+            if c == ' ' {
```
Comment (kimhanbeom) [inline src/encoding_decoding/decode_the_message.rs:65]: `get(c)`을 수행할때 값을 못 얻는 경우는 `  ` 뿐이므로 map내의 if문과 match 대신 `unwrap_or(&c)` 을 사용하면 코드가 더 간결해 질 것 같습니다.

**Review by kimhanbeom — CHANGES_REQUESTED** [review]

File: src/encoding_decoding/decode_the_message.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+    let decoded_message = message
+        .chars()
+        .map(|c| {
+            if c == ' ' {
+                ' '
+            } else {
+                *key_chars_map.get(&c).unwrap_or(&c)
```
Comment (kimhanbeom) [inline src/encoding_decoding/decode_the_message.rs:65]: c가 스페이스 공백일 경우에도 `unwrap_or` 이 수행될 것이므로 위의 `if`문도 제거해주세요

**Review by kimhanbeom — CHANGES_REQUESTED** [review]

File: src/encoding_decoding/decode_the_message.rs (hunk 5 above)
Comment (zmrdltl) [reply src/encoding_decoding/decode_the_message.rs]: 수정했습니다.

**Review by kimhanbeom — APPROVED** [review]
